CRAYSAT-2027: Add `sat bootprep` tests for resolve-branches behavior

annapoorna-s-alt opened the pull request:
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [issue id](issue link)
* Change will also be needed in `<insert branch name here>`
* Future work required by [issue id](issue link)
* Documentation changes required in [issue id](issue link)
* Merge with/before/after `<insert PR URL here>`

## Testing

_List the environments in which these changes were tested._

### Tested on:

  * `<development system>`
  * Local development environment
  * Virtual Shasta

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)?
- Were continuous integration tests run? If not, why?
- Was upgrade tested? If not, why?
- Was downgrade tested? If not, why?
- Were new tests (or test issues/Jiras) created for this change?

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_


## Pull Request Checklist

- [ ] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [ ] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

annapoorna-s-alt commented:
Tested on rocket
```
ncn-m001:/mnt/developer/sann/automation-test/bootprep/9-9/unstable # ./opt/cray/tests/install/ncn/scripts/python/sat_functional run bootprep.test_bootprep
test_configs_images_and_session_templates (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepConfigsImagesAndSessionTemplates)
Test creating, skipping and overwriting configurations, images and session templates ... ok
test_additional_inventory (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating CFS configurations with additional inventory ... ok
test_git_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations with git-based layers ... ok
test_missing_playbook (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with a missing playbook using CFS v2 succeeds ... ok
test_no_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that a file with an empty list of configs creates no configs ... ok
test_no_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with no layers ... ok
test_no_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations with git-based layers with --no-resolve-branches ... ok
test_product_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations using product-based layers ... ok
test_special_parameters (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with special parameters ... ok
test_additional_inventory (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating CFS configurations with additional inventory ... ok
test_git_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations with git-based layers ... ok
test_missing_playbook (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with a missing playbook using CFS v3 fails ... ok
test_no_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that a file with an empty list of configs creates no configs ... ok
test_no_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with no layers ... ok
test_no_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations with git-based layers with --no-resolve-branches ... ok
test_product_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations using product-based layers ... ok
test_special_parameters (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with special parameters ... ok
test_dry_run_and_save (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepDryRun)
Test running bootprep in dry-run and saving files ... ok
test_if_exists_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepIfExistsProperty)
Test the 'if_exists' property for CFS configurations ... ok
test_if_exists_images (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepIfExistsProperty)
Test the 'if_exists' property for IMS images ... ok
test_if_exists_session_templates (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepIfExistsProperty)
Test the 'if_exists' property for BOS session templates ... ok
test_image_customization (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepImageCustomizationCFSV2)
Test creating an ims image with a configuration ... ok
test_image_customization (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepImageCustomizationCFSV3)
Test creating an ims image with a configuration ... ok
test_image_customization_fail (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepImageCustomizationFailure)
Test creating a failing ims image ... ok
test_limit_option (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepLimitOption)
Test limit option when creating items. Use dry-run for images to speed up the test ... ok
test_session_template (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepSessionTemplates)
Test creating a bos session template ... ok

----------------------------------------------------------------------
Ran 26 tests in 2061.647s

OK

```

haasken-hpe commented on src/csm_testing/tests/sat_functional/bootprep/test_bootprep.py:
```diff
@@ -965,6 +965,17 @@ def test_git_layers(self):
+    @skip_test_if_csm_var_missing(['branch_name', 'commit_hash'])
+    def test_no_resolve_branches(self):
+        """Test creating multiple CFS configurations with git-based layers with --no-resolve-branches"""
+        result = self.run_bootprep('git-layers-config.yaml', '--format json', '--no-resolve-branches')
```
All the bootprep options are supposed to be in a single string in the second argument of this function. The third argument is `check`, which does not need to be changed from the default.

```suggestion
        result = self.run_bootprep('git-layers-config.yaml', '--format json --no-resolve-branches')
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/bootprep/test_bootprep.py:
```diff
@@ -965,6 +965,17 @@ def test_git_layers(self):
+        self.assertEqual(3, len(report['configurations']))
+
+        for config in report['configurations']:
+            self.validate_cfs_config(config['name'])
```
I would like to see a small amount of additional validation of the created CFS configurations here. When `--no-resolve-branches` is specified, the resulting configurations should still include the branch name in the layers if the layer in the bootprep file specified a branch. This is true for both CFS v3 and v2 although our methods of achieving this result are different for the two CFS versions.

There are three configurations created by the `git-layers-config.yaml` bootprep input file. The first, with name `{{test.prefix}}-csm-git-commit-layer`, has only one layer, and it does not specify a branch, so it should not have a branch in the created CFS configuration. The second, with name `{{test.prefix}}-csm-git-branch-layer`, also has only a single layer, and it does specify a branch, so it should have a branch in the created CFS configuration. The last configuration, with name `{{test.prefix}}-csm-two-git-layers`, has two layers, and the first does not specify a branch, while the second does. The created CFS configuration should have a branch in only the second layer.

This could be simplified a little bit if you created a simpler new bootprep input file that only creates a single configuration, something like:

```
---
# Define a configuration that defines a layer that specifies a git clone URL and branch
configurations:
- name: "{{test.config_name}}"
  layers:
  - name: csm-packages
    playbook: csm_packages.yml
    git:
      url: "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git"
      branch: "{{csm.branch_name}}"

```

Then you can just check for existence of the branch in the resulting created CFS configuration for a single configuration with a single layer. You could even modify the `validate_cfs_config` method to add this as an optional check, if you wanted.

Finally, although `test_git_layers` does exercise the `sat bootprep` code without `--no-resolve-branches`, it does not actually check that the branch names are dropped in the resulting configurations. I would add another test method that does explicitly check that. You could name it `test_resolve_branches` and use the same simplifiied bootprep input file from earlier in this comment. The test would be the same, but it would not pass `--no-resolve-branches`, and it would check that the branch does not exist in the resulting CFS configuration.

annapoorna-s-alt commented:
Testing output on rocket after addressing the comments:
```
ncn-m001:/mnt/developer/sann/automation-test/bootprep/15-9/final # ./opt/cray/tests/install/ncn/scripts/python/sat_functional run bootprep.test_bootprep
test_configs_images_and_session_templates (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepConfigsImagesAndSessionTemplates)
Test creating, skipping and overwriting configurations, images and session templates ... ok
test_additional_inventory (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating CFS configurations with additional inventory ... ok
test_git_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations with git-based layers ... ok
test_missing_playbook (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with a missing playbook using CFS v2 succeeds ... ok
test_no_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that a file with an empty list of configs creates no configs ... ok
test_no_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with no layers ... ok
test_no_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that branch is present in the layer when --no-resolve-branches is used. ... ok
test_product_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations using product-based layers ... ok
test_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that branch is NOT present in the layer when --no-resolve-branches is NOT used. ... ok
test_special_parameters (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with special parameters ... ok
test_additional_inventory (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating CFS configurations with additional inventory ... ok
test_git_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations with git-based layers ... ok
test_missing_playbook (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with a missing playbook using CFS v3 fails ... ok
test_no_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that a file with an empty list of configs creates no configs ... ok
test_no_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with no layers ... ok
test_no_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that branch is present in the layer when --no-resolve-branches is used. ... ok
test_product_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations using product-based layers ... ok
test_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that branch is NOT present in the layer when --no-resolve-branches is NOT used. ... ok
test_special_parameters (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with special parameters ... ok
test_dry_run_and_save (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepDryRun)
Test running bootprep in dry-run and saving files ... ok
test_if_exists_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepIfExistsProperty)
Test the 'if_exists' property for CFS configurations ... ok
test_if_exists_images (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepIfExistsProperty)
Test the 'if_exists' property for IMS images ... ok
test_if_exists_session_templates (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepIfExistsProperty)
Test the 'if_exists' property for BOS session templates ... ok
test_image_customization (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepImageCustomizationCFSV2)
Test creating an ims image with a configuration ... ok
test_image_customization (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepImageCustomizationCFSV3)
Test creating an ims image with a configuration ... ok
test_image_customization_fail (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepImageCustomizationFailure)
Test creating a failing ims image ... ok
test_limit_option (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepLimitOption)
Test limit option when creating items. Use dry-run for images to speed up the test ... ok
test_session_template (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepSessionTemplates)
Test creating a bos session template ... ok

----------------------------------------------------------------------
Ran 28 tests in 1993.291s

OK

```

haasken-hpe commented:
I ran the relevant tests again after my minor modifications:

```
ncn-m001:/mnt/developer/haasken/CRAYSAT-2027 # ./opt/cray/tests/install/ncn/scripts/python/sat_functional run bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2
test_additional_inventory (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating CFS configurations with additional inventory ... ok
test_git_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations with git-based layers ... ok
test_missing_playbook (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with a missing playbook using CFS v2 succeeds ... ok
test_no_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that a file with an empty list of configs creates no configs ... ok
test_no_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with no layers ... ok
test_no_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that branch is present in the layer when --no-resolve-branches is used. ... ok
test_product_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating multiple CFS configurations using product-based layers ... ok
test_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test that branch is NOT present in the layer when --no-resolve-branches is NOT used. ... ok
test_source_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating CFS configurations with source and clone url properties ... ok
test_special_parameters (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV2)
Test creating a CFS configuration with special parameters ... ok

----------------------------------------------------------------------
Ran 10 tests in 169.147s

OK
ncn-m001:/mnt/developer/haasken/CRAYSAT-2027 # ./opt/cray/tests/install/ncn/scripts/python/sat_functional run bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3
test_additional_inventory (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating CFS configurations with additional inventory ... ok
test_git_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations with git-based layers ... ok
test_missing_playbook (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with a missing playbook using CFS v3 fails ... ok
test_no_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that a file with an empty list of configs creates no configs ... ok
test_no_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with no layers ... ok
test_no_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that branch is present in the layer when --no-resolve-branches is used. ... ok
test_product_layers (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating multiple CFS configurations using product-based layers ... ok
test_resolve_branches (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test that branch is NOT present in the layer when --no-resolve-branches is NOT used. ... ok
test_source_configs (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating CFS configurations with source and clone url properties ... ok
test_special_parameters (csm_testing.tests.sat_functional.bootprep.test_bootprep.TestBootprepCreateConfigsCFSV3)
Test creating a CFS configuration with special parameters ... ok

----------------------------------------------------------------------
Ran 10 tests in 169.454s

OK
```
